File: Breadth-First Search/Binary tree/LET_no993_Cousins in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        int parent;
        queue<pair<TreeNode*, TreeNode*>> q;
        q.push(make_pair(nullptr, root));
        
        while (!q.empty()) {
            parent = 0;
            for (int i = q.size(); i > 0; i--) {
                auto cur = q.front(); q.pop();
                if (cur.first && (cur.second->val == x || cur.second->val == y)) {
                    if (!parent) 
                        parent = cur.first->val;
                    else 
                        return parent != cur.first->val;
                }
                
                if (cur.second->left) 
                    q.push(make_pair(cur.second, cur.second->left));
                if (cur.second->right) 
                    q.push(make_pair(cur.second, cur.second->right));
            }
        }
        return false;
    }
};

File: Breadth-First Search/Binary tree/LET_no993_Cousins in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        queue = collections.deque([(None, root)])
        while queue:
            parent = 0
            for _ in range(len(queue)):
                p, c = queue.popleft()
                if p and (c.val == x or c.val == y):
                    if not parent: parent = p.val
                    else: return parent != p.val
                if c.left: queue.append((c, c.left))
                if c.right: queue.append((c, c.right))
        return False